frame/vesting/src/lib.rs

```diff
@@ -312,105 +369,382 @@ pub mod pallet {
+			);
+			Self::write_vesting(&who, schedules)?;
+			Self::write_lock(&who, locked_now);
 
 			Ok(())
 		}
 	}
 }
 
 impl<T: Config> Pallet<T> {
-	/// (Re)set or remove the pallet's currency lock on `who`'s account in accordance with their
-	/// current unvested amount.
-	fn update_lock(who: T::AccountId) -> DispatchResult {
-		let vesting = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+	// Create a new `VestingInfo`, based off of two other `VestingInfo`s.
+	// NOTE: We assume both schedules have had funds unlocked up through the current block.
+	fn merge_vesting_info(
+		now: T::BlockNumber,
+		schedule1: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+		schedule2: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+	) -> Result<Option<VestingInfo<BalanceOf<T>, T::BlockNumber>>, DispatchError> {
+		let schedule1_ending_block = schedule1.ending_block_as_balance::<T::BlockNumberToBalance, T>()?;
+		let schedule2_ending_block = schedule2.ending_block_as_balance::<T::BlockNumberToBalance, T>()?;
+		let now_as_balance = T::BlockNumberToBalance::convert(now);
+
+		// Check if one or both schedules have ended.
+		match (schedule1_ending_block <= now_as_balance, schedule2_ending_block <= now_as_balance) {
+			// If both schedules have ended, we don't merge and exit early.
+			(true, true) => return Ok(None),
+			// If one schedule has ended, we treat the one that has not ended as the new
+			// merged schedule.
+			(true, false) => return Ok(Some(schedule2)),
+			(false, true) => return Ok(Some(schedule1)),
+			// If neither schedule has ended don't exit early.
+			_ => {}
+		}
+
+		let locked = schedule1
+			.locked_at::<T::BlockNumberToBalance>(now)
+			.saturating_add(schedule2.locked_at::<T::BlockNumberToBalance>(now));
+		// This shouldn't happen because we know at least one ending block is greater than now.
+		debug_assert!(!locked.is_zero(), "merge_vesting_info validation checks failed to catch a locked of 0");
+		if locked.is_zero() {
+			log::warn!(
+				target: LOG_TARGET,
+				"merge_vesting_info validation checks failed to catch a locked of 0"
+			);
+			return Ok(None);
+		}
+
+		let ending_block = schedule1_ending_block.max(schedule2_ending_block);
+		let starting_block = now.max(schedule1.starting_block()).max(schedule2.starting_block());
+		let duration =
+			ending_block.saturating_sub(T::BlockNumberToBalance::convert(starting_block));
+		let per_block = if duration > locked {
+			// This would mean we have a per_block of less than 1, which should not be not possible
+			// because when we create the new schedule is at most the same duration as the longest,
+			// but never greater.
+			One::one()
+		} else {
+			match locked.checked_div(&duration) {
+				// The logic of `ending_block_as_balance` guarantees that each schedule ends at
+				// least a block after it starts and since we take the max starting and ending_block
+				// we should never get here.
+				None => locked,
+				Some(per_block) => per_block,
+			}
```

Comment:
this whole thing looks like it can be simplified to:

```suggestion
		let duration =
			ending_block.saturating_sub(T::BlockNumberToBalance::convert(starting_block)).max(One::one());
		let per_block = (locked / duration).max(One::one());
```

Reply:
Nice, I implemented here: c750bd8b2f09da871a2f87c0a11cc4f92a31cc59